src/Razor/src/Microsoft.AspNetCore.Razor.ExternalAccess.RoslynWorkspace/RazorWorkspaceListenerBase.cs

```diff
@@ -272,22 +286,123 @@ private static async Task ProcessWorkCoreAsync(ImmutableArray<Work> work, Stream
+        var intermediateOutputPath = Path.GetDirectoryName(project.CompilationOutputInfo.AssemblyPath);
+        if (intermediateOutputPath is null)
+        {
+            _logger.LogInformation("intermediateOutputPath is null, skip update for {projectId}", project.Id);
+            return;
+        }
+
+        if (!_projectEntryMap.TryGetValue(project.Id, out var entry))
+        {
+            _projectEntryMap[project.Id] = entry = new();
+        }
+
+        var (didChange, projectInfo) = await DidProjectInfoChangeAsync(project, entry, projectPath, intermediateOutputPath, cancellationToken).ConfigureAwait(false);
+        if (didChange)
+        {
+            projectInfo.AssumeNotNull();
+            UpdateEntry(project, projectInfo, entry);
+
+            stream.WriteProjectInfoAction(ProjectInfoAction.Update);
+            await stream.WriteProjectInfoAsync(projectInfo, cancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    private void UpdateEntry(Project project, RazorProjectInfo projectInfo, ProjectEntry entry)
+    {
+        entry.RootNamespace = projectInfo.RootNamespace;
+        entry.ConfigurationHash = projectInfo.Configuration.GetHashCode();
+        entry.DocumentsHash = projectInfo.Documents.GetHashCode();
+
+        _cachedTagHelperResolver.TryGetId(project.Id, out var resultId);
+        entry.TagHelpersResultId = resultId;
+    }
+
+    private async Task<(bool, RazorProjectInfo? projectInfo)> DidProjectInfoChangeAsync(Project project, ProjectEntry entry, string projectPath, string intermediateOutputPath, CancellationToken cancellationToken)
+    {
+        ImmutableArray<DocumentSnapshotHandle> documents = default;
+        var csharpLanguageVersion = (project.ParseOptions as CSharpParseOptions)?.LanguageVersion ?? LanguageVersion.Default;
+        CachedTagHelperResolver.DeltaResult? delta = null;
+
+        var configuration = RazorProjectInfoHelpers.ComputeRazorConfigurationOptions(project.AnalyzerOptions.AnalyzerConfigOptionsProvider, out var rootNamespace);
+        if (configuration is null)
+        {
+            return (false, null);
+        }
+
+        if (entry.RootNamespace is null)
+        {
+            return await GetChangedValueAsync().ConfigureAwait(false);
+        }
+
+        if (!entry.RootNamespace.Equals(rootNamespace))
+        {
+            return await GetChangedValueAsync().ConfigureAwait(false);
+        }
+
+        if (entry.ConfigurationHash != configuration.GetHashCode())
+        {
+            return await GetChangedValueAsync().ConfigureAwait(false);
+        }
+
+        documents = RazorProjectInfoHelpers.GetDocuments(project, projectPath);
+        if (documents.Length == 0)
+        {
+            _logger.LogInformation("No razor documents in {projectId}", project.Id);
+            return (false, null);
+        }
+
+        if (entry.DocumentsHash != documents.GetHashCode())
```

Comment:
This won't do what you want. Instead, it'll likely return true almost every time. `RazorProjectInfoHelpers.GetDocumts()` returns a new `ImmutableArray<DocumentSnapshotHandle>` each time it's called, and `ImmutableArray<T>.GetHashCode()` isn't computed from the contents; it just passes the reference to the underlying array.

To address this, please use `Microsoft.AspNetCore.Razor.Utilities.Checksum` and build up a checksum.